skimage/future/graph/graph_cut.py

```diff
@@ -258,37 +325,67 @@ def _ncut_relabel(rag, thresh, num_cuts):
         The array which maps old labels to new ones. This is modified inside
         the function.
     """
+    thresh = init_thresh
     d, w = _ncut.DW_matrices(rag)
     m = w.shape[0]
 
+    # If 2 regions not cutting is optimal choice
     if m > 2:
         d2 = d.copy()
         # Since d is diagonal, we can directly operate on its data
         # the inverse of the square root
         d2.data = np.reciprocal(np.sqrt(d2.data, out=d2.data), out=d2.data)
 
-        # Refer Shi & Malik 2001, Equation 7, Page 891
-        vals, vectors = linalg.eigsh(d2 * (d - w) * d2, which='SM',
-                                     k=min(100, m - 2))
-
-        # Pick second smallest eigenvector.
-        # Refer Shi & Malik 2001, Section 3.2.3, Page 893
-        vals, vectors = np.real(vals), np.real(vectors)
-        index2 = _ncut_cy.argmin2(vals)
-        ev = vectors[:, index2]
+        # Refer Shi & Malik 2000, Equation 7, Page 891
+        v0 = 2 * np.random.random((m,)) - 1
+        v0 /= np.linalg.norm(v0)
+        vals, vectors = eigsh(d2 * (d - w) * d2, which='SM',
+                              k=min(m-1, 100), v0=v0)
+
+        # Build vector orthogonal to d from first two eigenvectors
+        # Helps avoid problems from degeneracy
+        # Refer Shi & Malik 2000, Section 2.1.10, Page 891
+        # Refer Shi & Malik 2000, Section 3.2.3, Page 893
+        vectors = (vectors[:, :2].T / d2.data).T
+        data_normed = d.data / np.dot(d.data, d.data)
+        ev = ((vectors[:, 0] - np.dot(vectors[:, 0], d.data) * data_normed) +
+              (vectors[:, 1] - np.dot(vectors[:, 1], d.data) * data_normed))
 
         cut_mask, mcut = get_min_ncut(ev, d, w, num_cuts)
-        if (mcut < thresh):
-            # Sub divide and perform N-cut again
-            # Refer Shi & Malik 2001, Section 3.2.5, Page 893
-            sub1, sub2 = partition_by_cut(cut_mask, rag)
-
-            _ncut_relabel(sub1, thresh, num_cuts)
-            _ncut_relabel(sub2, thresh, num_cuts)
-            return
-
-    # The N-cut wasn't small enough, or could not be computed.
-    # The remaining graph is a region.
-    # Assign `ncut label` by picking any label from the existing nodes, since
-    # `labels` are unique, `new_label` is also unique.
-    _label_all(rag, 'ncut label')
+
+        # These variables no longer needed
+        del d, d2, w, vals, vectors, ev
```

Comment:
Do these `del` calls strictly affect performance? I would rather leave memory management to the interpreter if possible...

Reply:
The del calls might be premature optimization. `vectors` can get up to 100 x N_{super pixels} and the code generates two `vectors` arrays every time it subdivides a graph. It appears in the body of a generator, so I don't think the interpreter would know it can be deallocated. 

I could put everything after the if statement into a function that returns `mcut` and `cut_mask` and remove the first `del`. The second `del` isn't as important since `cut_mask` is smaller, but I don't think `cut_mask` can be treated the same way.